Autotools build types

pretyman opened the pull request:
Changes discussed in https://github.com/ruslo/hunter/issues/411

ruslo commented on cmake/modules/hunter_autotools_project.cmake:
```diff
@@ -94,7 +99,17 @@ function(hunter_autotools_project target_name)
   hunter_test_string_not_empty("${PARAM_INSTALL_DIR}")
+  hunter_test_string_not_empty("${PARAM_PACKAGE_CONFIGURATION_TYPES}")
 
+  list(LENGTH PARAM_PACKAGE_CONFIGURATION_TYPES _len)
```
We are inside the function so it's okay to use "normal" variable names without "_" at start (see [Naming](http://0.readthedocs.io/en/latest/cmake.html#naming)).

ruslo commented on cmake/projects/Avahi/hunter.cmake:
```diff
@@ -20,10 +21,11 @@ hunter_add_version(
 
+hunter_configuration_types(Avahi CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_avahi_autotools)
 hunter_download(PACKAGE_NAME Avahi)
```
I'm afraid that since internal structure of scheme changed we have to add `PACKAGE_INTERNAL_DEPS_ID`. See https://github.com/ruslo/hunter/wiki/dev.modules#hunter_download. Problem will occur in case somebody already save this package in cache by adding `HUNTER_CONFIGURATION_TYPES=Release`. It means if somebody build all 3rd parties with `Release` then `Avahi` saved to cache without new flags. Now when we add flags cache system doesn't see updates because think that: source doesn't change + build type doesn't change (Release was global, now it's local default but still Release too) + internal deps id doesn't change => using version from cache (without flags).

pretyman replied:
Oh, I see. So adding `PACKAGE_INTERNAL_DEPS_ID "1"` should be enough, correct?

ruslo replied:
> Oh, I see. So adding PACKAGE_INTERNAL_DEPS_ID "1" should be enough, correct?

yes

> I ask to make sure because there are quite many files to update

Yes, it's true. Also there are a lot of packages to test.

pretyman replied:
> Yes, it's true. Also there are a lot of packages to test.

Any ideas how to fire ingenue for the packages that were changed?

ruslo replied:
> Any ideas how to fire ingenue for the packages that were changed?

@pretyman need to write helper script (or better a service) :smile:

ruslo replied:
Need `PACKAGE_INTERNAL_DEPS_ID` here

pretyman commented:
Updated this pull request: fixed the naming of variable in cmake/modules/hunter_autotools_project.cmake and bumped PACKAGE_INTERNAL_DEPS_ID for cacheable packages

ruslo commented on cmake/modules/hunter_autotools_project.cmake:
```diff
@@ -159,7 +174,9 @@ function(hunter_autotools_project target_name)
-  get_directory_property(defs COMPILE_DEFINITIONS)
+  get_directory_property(defs
+      COMPILE_DEFINITIONS_${_config_type}
+  )
```
I think we should use both `COMPILE_DEFINITIONS` and `COMPILE_DEFINITIONS_<CONFIG>`

ruslo commented on cmake/modules/hunter_autotools_project.cmake:
```diff
@@ -178,21 +195,21 @@ function(hunter_autotools_project target_name)
   #
   # C Compiler Flags (defines or include directories should not be needed here)
-  set(cflags "${CMAKE_C_FLAGS} ${PARAM_CFLAGS}")
+  set(cflags "${CMAKE_C_FLAGS_${_config_type}} ${PARAM_CFLAGS}")
```
same as above

ruslo commented on cmake/modules/hunter_autotools_project.cmake:
```diff
@@ -178,21 +195,21 @@ function(hunter_autotools_project target_name)
   # CXXFLAGS=${cxxflags} ${CMAKE_CXX_FLAGS}
   #
   # C++ Compiler flags (defines or include directories should not be needed here)
-  set(cxxflags "${CMAKE_CXX_FLAGS} ${PARAM_CXXFLAGS}")
```
same as above

ruslo commented on cmake/modules/hunter_autotools_project.cmake:
```diff
@@ -178,21 +195,21 @@ function(hunter_autotools_project target_name)
   # LDFLAGS=${ldflags}
   #
   # Linker flags
-  set(ldflags "${CMAKE_EXE_LINKER_FLAGS} ${PARAM_LDFLAGS}")
```
same as above

pretyman commented:
My god, thanks for the review, I was going to cause some real damage by omitting the regular compiling and linking flags.

Updated the code to rebase onto develop and include the autogen.sh changes.

[sorry for taking so long, ran into some trouble]

ruslo commented on cmake/projects/PostgreSQL/hunter.cmake:
```diff
@@ -18,6 +19,7 @@ hunter_add_version(
 
+hunter_configuration_types(PostgreSQL CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_download(PACKAGE_NAME PostgreSQL)
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

pretyman replied:
This package is not cacheable, do I still need to have the `PACKAGE_INTERNAL_DEPS_ID` for it? Maybe something changed, but I thought I had to add to cacheable packages only

ruslo replied:
> This package is not cacheable

Okay, not needed in this case.

ruslo commented on cmake/projects/intltool/hunter.cmake:
```diff
@@ -18,5 +19,6 @@ hunter_add_version(
 
+hunter_configuration_types(intltool CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_download(PACKAGE_NAME intltool)
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

ruslo commented on cmake/projects/libdaemon/hunter.cmake:
```diff
@@ -18,5 +19,6 @@ hunter_add_version(
 
+hunter_configuration_types(libdaemon CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_download(PACKAGE_NAME libdaemon)
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

ruslo commented on cmake/projects/odb-boost/hunter.cmake:
```diff
@@ -18,6 +19,7 @@ hunter_add_version(
 
+hunter_configuration_types(odb-boost CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_odb-boost_autotools)
 hunter_download(PACKAGE_NAME odb-boost)
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

ruslo commented on cmake/projects/odb-mysql/hunter.cmake:
```diff
@@ -18,6 +19,7 @@ hunter_add_version(
 
+hunter_configuration_types(odb-mysql CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_odb-mysql_autotools)
 hunter_download(PACKAGE_NAME odb-mysql)
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

ruslo commented on cmake/projects/odb-pgsql/hunter.cmake:
```diff
@@ -18,6 +19,7 @@ hunter_add_version(
 
+hunter_configuration_types(odb-pgsql CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_odb-pgsql_autotools)
 hunter_download(PACKAGE_NAME odb-pgsql)
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

ruslo commented on cmake/projects/odb-sqlite/hunter.cmake:
```diff
@@ -21,6 +22,7 @@ hunter_add_version(
 
+hunter_configuration_types(odb-sqlite CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_odb-sqlite_autotools)
 hunter_download(
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

ruslo replied:
Also I guess I missed that before: probably there is no need for `PACKAGE_DEPENDS_ON`

pretyman replied:
ok, will remove in a separate commit to not mix things up

ruslo commented on cmake/projects/odb/hunter.cmake:
```diff
@@ -18,5 +19,6 @@ hunter_add_version(
 
+hunter_configuration_types(odb CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_download(PACKAGE_NAME odb)
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

ruslo commented on cmake/projects/presentproto/hunter.cmake:
```diff
@@ -20,6 +21,7 @@ hunter_add_version(
+hunter_configuration_types(presentproto CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_cacheable(presentproto)
 hunter_download(
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

pretyman replied:
thanks

ruslo commented on cmake/projects/pthread-stubs/hunter.cmake:
```diff
@@ -20,6 +21,7 @@ hunter_add_version(
+hunter_configuration_types(pthread-stubs CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_cacheable(pthread-stubs)
 hunter_download(
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

pretyman replied:
thanks

ruslo commented on cmake/projects/xcb-proto/hunter.cmake:
```diff
@@ -32,6 +33,7 @@ hunter_add_version(
+hunter_configuration_types(xcb-proto CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_cacheable(xcb-proto)
 hunter_download(
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

pretyman replied:
thanks

ruslo commented on cmake/projects/xextproto/hunter.cmake:
```diff
@@ -20,6 +21,7 @@ hunter_add_version(
+hunter_configuration_types(xextproto CONFIGURATION_TYPES Release)
 hunter_pick_scheme(DEFAULT url_sha1_autotools)
 hunter_cacheable(xextproto)
 hunter_download(
```
Need `PACKAGE_INTERNAL_DEPS_ID` update

pretyman replied:
thanks

pretyman commented:
updated projects `presentproto pthread-stubs xcb-proto xextproto` which were cacheable and did not have `PACKAGE_INTERNAL_DEPS_ID`

ruslo commented:
I think `hunter_configuration_types(... CONFIGURATION_TYPES Release)` is missing for `libsodium` and `geos`, `PACKAGE_INTERNAL_DEPS_ID "1"` missing for libsodium. Probably because this is new packages (?)

pretyman commented:
updated `geos` and `libsodium`, thanks for the heads up

ruslo commented on cmake/projects/geos/hunter.cmake:
```diff
@@ -1,9 +1,14 @@
 # Copyright (c) 2015, Ruslan Baratov, Alexandre Pretyman
 # All rights reserved.
 
-# !!! DO NOT PLACE HEADER GUARDS HERE !!!
```
Keep this line! :)

pretyman replied:
ooops, sorry. I don't remember touching it, something must have gone wrong on my side

ruslo commented on cmake/projects/libsodium/hunter.cmake:
```diff
@@ -17,10 +18,14 @@ hunter_add_version(
 
+hunter_configuration_types(libsodium CONFIGURATION_TYPES Release)
 if(MSVC)
     hunter_pick_scheme(DEFAULT url_sha1_libsodium_msbuild)
```
This scheme works fine with build types: https://github.com/ruslo/hunter/blob/master/cmake/projects/libsodium/schemes/url_sha1_libsodium_msbuild.cmake.in

ruslo commented on cmake/projects/libsodium/hunter.cmake:
```diff
@@ -17,10 +18,14 @@ hunter_add_version(
 if(MSVC)
     hunter_pick_scheme(DEFAULT url_sha1_libsodium_msbuild)
 else()
     hunter_pick_scheme(DEFAULT url_sha1_libsodium_autogen_autotools)
```
I see different build types but flags not passed to the build. @Cyberunner23 does Release/Debug specific flags used in `url_sha1_libsodium_autogen_autotools` scheme?

pretyman replied:
Let me investigate this, somewhere along this change I did a `git clean`, so I ended up losing some things that could have caused this.

pretyman replied:
Its a C library only, I see that the `CFLAGS` are being passed and being used. C++ flags under `CXXFLAGS` are not used. What kind of flags were you expecting? Maybe my test is wrong :|

Cyberunner23 replied:
> I see different build types but flags not passed to the build. @Cyberunner23 does Release/Debug specific flags used in url_sha1_libsodium_autogen_autotools scheme?

There are no specific flags for each build type, in fact it seems that it always produces debug builds

pretyman replied:
> in fact it seems that it always produces debug builds no matter what option I set

That's what is this pull request hoping to fix :>

Cyberunner23 replied:
Just made an issue upstream about this, currently it seems there is no way to make a Release build with libsodium

Cyberunner23 replied:
got a response `env CFLAGS=-O2 ./configure ...` is the way to make a release build

Cyberunner23 replied:
want me to change it to include this?

ruslo replied:
> want me to change it to include this?

@Cyberunner23 Please take a look at this pull request I think you just need to use newly introduced feature.

@pretyman Okay, since this affect only one package with separate build scheme, let's ignore it for now.

Cyberunner23 replied:
Just tested it, seems like symbold are still present, might need to manually call strip on the library files

pretyman replied:
I was not able to use `make VERBOSE=1` to print the full compilation command line during the build for some reason, even with `make VERBOSE=1` I get the terse compilation output as something along the lines of `CC dir/file.c`

Cyberunner23 replied:
use ./configure --disable-silent-rules to get the full compilation command line

Cyberunner23 replied:
using `env CFLAGS=-O2` still produces symbols but no debugging symbols, checked with gdb

pretyman commented:
Updated: removed the include guards of `geos`

ruslo commented:
SQLite3 testing:
* https://travis-ci.org/ingenue/hunter/builds/145337943

ruslo commented:
Looks good:
```
/bin/bash ./libtool ... -O3 -DNDEBUG ... -c -o sqlite3.lo sqlite3.c
```

```
... CC=/usr/bin/gcc-4.8 CXX=/usr/bin/g++-4.8 "CFLAGS=-O3 -DNDEBUG" "CXXFLAGS=-O3 -DNDEBUG -std=c++11" ...
```

ruslo commented:
Released: https://github.com/ruslo/hunter/releases/tag/v0.14.31

pretyman commented:
Thanks! 👍
